### ISSUE TITLE
Autotype is opening main Keepass window and sending keypresses to the wrong window

### ISSUE THREAD

**Issue by cumber** [keepassxreboot/keepassxc#8169]
## Overview

Since upgrading to 2.7.1 a few days ago (from 2.6.6), I've been having troubles with autotype. It seems to bring up the main KeepassXC window and then send the autotyped keypresses to Keepass instead of whatever application I was trying to autotype my credentials into

I'm using a Nixos Linux system with XMonad, which is a tiling window manager; the exact behaviour of the new Keepass window gaining focus is probably affected by this, but the core issue of the main window being opened at the wrong time is more likely an issue in KeepassXC itself?

## Steps to Reproduce
 
1. Make sure KeepassXC is running in the tray, with no windows open
2. Open a web browser, go to a site where you want to log in.
3. Hit the autotype activation key
4. Double click an entry, press enter, or hit the `Type Sequence` button

## Expected Behavior

I expect the autotype selection dialogue to close, and the autotype sequence of keypresses to be delivered to the application that had focus before it opened (the browser in my example).

This is exactly the behaviour I've had with KeepassXC 2.6.6 (and earlier versions).

## Actual Behavior

The autotype selection dialogue closes, but the main KeepassXC window opens unexpectedly. This new window gains focus (at least on my system), so it receives the autotype sequence of keypresses (which has the potential to make destructive changes to my password database if the exact wrong sequence of keys happens to be in the credentials being autotyped!)

If the main window is already open, it does not gain focus and autotype works as expected. So leaving it open on a spare workspace is a workaround.

## Context

KeePassXC - Version 2.7.1

Qt 5.15.3
Debugging mode is disabled.

Operating system: NixOS 22.11 (Raccoon)
CPU architecture: x86_64
Kernel: linux 5.15.46

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.18.1

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux (NixOS)
Desktop Env: XMonad
Windowing System: X11

KeepassXC is installed from nixpkgs revision 6616de389ed55fba6eeba60377fc04732d5a207c, if you need to trace the exact build

**Comment by cumber** [comment]
This is not a bug?

**Comment by cumber** [comment]
With the release of 2.7.4 this problem is no longer occurring. Not sure whether it was deliberately addressed or just accidentally fixed as a side effect of some other change, but either way thanks!

**Comment by cumber** [comment]
It seems I was mistaken; this issue is still affecting me in 2.7.4 after all.

**Comment by stefan123t** [comment]
For me it this is the precise sequence:

* browser with login form
* press auto-type hotkey (Ctrl+Alt+A)
* KeePassXC main window of *`<Database name> - KeepassXC`* **and** 
  * a) the *`Unlock KeePassXC Database`* dialog (optional, first)
  * b) the *`Auto-Type - KeePassXC`* dialog 
  **pop-up almost simultaneously**.
* when I click into the *`KeePassXC - Auto-Type`* (or *`Unlock KeePassXC Database`*) dialog the main window comes to the foreground too.

Before in KeePassXC 2.6 and other KeePass implementations the *`Auto-Type - KeePassXC`* (and *`Unlock KeePassXC Database`*) dialogs were somehow not linked together to the main window.

I am also on 2.7.4 for a while now, running under Linux Mint and Cinnamon.
Though auto-type does work in general with multiple keyboard layouts:
Many thanks :clap: to @hifi and @droidmonkey time & again!

But sometimes I need to double-check, which Password Auto-Type entry I have to select, because some of the login dialogs have my username stored and some don't. I have therefor separate entries for *`Password only`* and *`Username & Password`*.
Now when I select the Auto-Type window (in order to move it slightly out of sight on the login form) the bigger / maximized Main Window also comes to the front and obstructs e.g. the Browser window. 

Also sometimes I want to verify whether the cursor is in the right form field before hitting Auto-Type.
So I have to switch back & forth between the Browser and Auto-Type window using Alt+Tab or cancel the Auto-Type again.
I kind of find it annoying to have the **main window brought up together with the Auto-Type sub-window/dialog**. 

I believe I have discussed this in the past with the developers and it may be related to how KDE (Qt / Plasma) works under Linux.
But I would also appreciate :+1: if the Auto-Type (and Unlock) windows could possibly be implemented as **unlinked windows / dialogs from the main window** again, maybe using a custom settings option ?

**Comment by droidmonkey** [comment]
Please provide full system details. There is no reason why the main window is appearing on Linux.  There is also no reason why the unlock dialog AND Auto-Type dialog appear together. I wonder if you have a corrupt installation that has old libraries still in place?

**Comment by stefan123t** [comment]
@droidmonkey I believe(d) this is still the same issue as in #6497
I have since updated KeePassXC to 2.7.4 but I can verify the other system specs if needed ?
It is indeed an ageing installation which I have since maintained for many years :smiley_cat:

```bash
KeePassXC - Version 2.7.4
Revision: 63b2394
Distribution: AppImage

Qt 5.15.2
Debugging mode is disabled.

Operating system: Linux Mint 20.3
CPU architecture: x86_64
Kernel: linux 5.15.0-60-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.19.1
```

I was annoyed once again and thought I check the current issues, whether this has been fixed or witnessed by / bothered someone else since back then. 
Apparently the coupling of the dialog with the main window (in Plasma?) was witnessed / reproduced by @hifi too and the change ìn #6497 has been delayed to 2.8.0 in the mean time.

That is the unlock dialog appears first with the main window and after that the auto-type dialog with the main window.
I just observed that both dialogs feature the same behaviour, when clicked into or on the dialog the main window will be brought to the front too.

**Comment by cumber** [comment]
@stefan123t This sounds like a different issue to me, although possibly it's related.

For me, the main window isn't brought up when autotype dialogs are *opened*. The main window only appears specifically when the actual autotype sequence is performed, after the dialog is *closed*; it pops up just in time to receive the autotype input instead of wherever it was supposed to go. And it only happens if the main window was minimised to the tray. Clicking in the autotype or unlock dialog doesn't bring the main window to the front.

**Comment by droidmonkey** [comment]
I cannot reproduce the issues identified in this thread on Fedora with Plasma desktop. I could reproduce the problem with the auto-type dialog not appearing on the current virtual desktop and also, as a result, not being able to type into applications on that virtual desktop.

To recap, I cannot (at all) reproduce the problem of the keepassxc main window appearing AFTER selecting the auto-type sequence.

**Comment by cumber** [comment]
So it's probably to do with Xmonad then (or at least a tiling window manager?). But the behaviour is at least *affected* by KeePass code, since it doesn't happen in 2.6 and does in 2.7.

I'm a developer, so I can try and debug this myself, I don't have a lot of C++ under my belt (and no Qt at all)...

Is there any sort of debug build I can run that would give you more information?

**Comment by droidmonkey** [comment]
You'll have to set breakpoints in various locations, I really can't walk you through that since it's complex and I haven't the time unfortunately.

**Comment by droidmonkey** [comment]
You'll have to set breakpoints in various locations, I really can't walk you through that since it's complex and I haven't the time unfortunately.

**Comment by WhiteWinterWolf** [comment]
I think I have the same problem here.

I have the impression that the issue occurs when these conditions are fulfilled:

- Auto-type is invoked while focus is on the targeted window using a global keyboard shortcut
- The KeepassXC database is currently locked
- The *"Minimize after unlock"* setting is enabled

In this situation, once the database has been unlocked, instead of typing the credential in the N-1 window, credentials are typed in the N-2 windows, meaning the window right below the expected window.

I have the impression that the Autotype feature doesn't take into account the *"Minimize after unlock"* setting: once the database has been unlocked, Auto-type expects the currently focused windows to be KeepassXC main window, so it switches the focus to the window right below and send its key presses.

However, when the *"Minimize after unlock"* setting is enabled, once the database has been unlocked, the currently focused window is already the one awaiting for the credentials. Autotype however still applies the same process as above: it switches the focus to the window right below which is completely unrelated and send the key presses to it.

In case of bad luck (think chat window for instance) this unrelated window may cause the credentials to be revealed to an unauthorized third-party :(

**Comment by droidmonkey** [comment]
@WhiteWinterWolf we used to have the behavior you described, but I recoded that for 2.7.0. Now the behavior is to capture the window you are currently focused on, perform whatever needs to happen in KeePassXC, then explicitly focus on the window you were on before. The code for that is here:
https://github.com/keepassxreboot/keepassxc/blob/27c5c5d464bb0140cf73b5954a4ad01de790e6ef/src/autotype/AutoType.cpp#L302-L305

We also make sure the window you intended to do auto-type on is still the one in focus before every key press:
https://github.com/keepassxreboot/keepassxc/blob/27c5c5d464bb0140cf73b5954a4ad01de790e6ef/src/autotype/AutoType.cpp#L320-L325

I couldn't reproduce your problem as described.

**Comment by WhiteWinterWolf** [comment]
@droidmonkey : Indeed. After digging a bit, I have the impression that, on my side at least, the issue is caused by windows which handle the focus in custom ways. In fact, I can now reproduce the issue at will using Yakuake, a dropdown terminal. Its window is special as it doesn't appear in the `Alt-Tab` windows list, but it nevertheless has a window title that KeepassXC Autotype can match.

As long as the KeepassXC database is already opened, I can use the Autotype global shortcut and have the keystrokes correctly sent to the Yakuake terminal with no issue. The same way, as long as the *"Minimize after unlock"* is disabled, I don't encounter the issue anymore.

However, I can reproduce the issue as follow:

1. Close KeepassXC application
2. Start KeepassXC application, but leave it on the database unlock screen, do not unlock it
3. Focus on some random application window, like a text editor
4. Focus on the dropdown terminal
5. Press the Global Autotype shortcut
6. KeepassXC gets the focus and asks to unlock the database: unlock it
7. KeepassXC keeps the focus and asks to select an auto-type sequence: choose one and validate
8. The focus now goes back to the random application from step 3) which receives KeepassXC keystrokes in place of the dropdown terminal.

I'm using KeepassXC version 2.7.4 on Debian Bookworm.

**Comment by droidmonkey** [comment]
I'm fairly certain this isn't actually happening: "Focus on the dropdown terminal"

I think you may think it is focused, but in reality, the notepad application still has the focus. That is the only explanation I can think of besides the Window pointer being null, in which case the global Auto-Type will act like the entry level Auto-Type and select the previously selected window (not by pointer). This could happen due to the way you mention this terminal working and not actually being part of the alt+tab menu.

**Comment by cumber** [comment]
Well I'm glad it's not only me that's having issues in this area!

The behaviour I see doesn't match what @WhiteWinterWolf describes though; I've *never* seen it focus on a **different** application. It always brings up the main KeepassXC window and focuses on that; despite there being this code:

https://github.com/keepassxreboot/keepassxc/blob/27c5c5d464bb0140cf73b5954a4ad01de790e6ef/src/autotype/AutoType.cpp#L388-L392

It also makes no difference whether the database was already locked or unlocked. If the database was locked, it'll bring up the unlock dialog, then the entry-selection dialog (all without the main KeepassXC window being visible), then the main window pops up and receives the auto-type input.

I also have the "Minimize after unlock" setting *disabled* already; interestingly when I *enable* it, the Keepass window still pops up but no it longer receives focus (the autotype actually goes to the correct window); but it doesn't actually minimize Keepass window.

**Comment by cumber** [comment]
@droidmonkey You mentioned the *old* logic that WhiteWinterWolf outlined was replaced in version 2.7, which lines up exactly with when this started happening for me. I think that new logic must make some assumptions that are not true on my system.

You mentioned the `window` pointer being used to identify the target window; how is that set? I tried tracing it back through GitHub code searches, but I got lost (it seemed to be `m_windowForGlobal`, but I don't know where that comes from). It seems like from the behaviour I observe that the window pointer is set to KeepassXC itself? Is it possible that the KeepassXC application is receiving focus (even though there are no open windows) in order to receive the hotkey event, and then setting the window pointer to KeepassXC?

How does registering a global hotkey work? If it's a standard API where the window manager has to be told to listen for the key combination and notify KeepassXC, then that could explain where the difference in my setup is (XMonad implementing that process differently than other more common window managers?)
